lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUTransformCharFilter.java

```diff
@@ -0,0 +1,384 @@
+   */
+  private boolean ensureRollbackBufferCapacity() {
+    // ensure space for at least 2 chars (surrogate pair, max possible space needed)
+    if (rollbackBuffer.length - rollbackBufferSize < 2) {
+      if (rollbackBuffer.length < maxRollbackBufferCapacity) {
+        rollbackBuffer = ArrayUtil.growExact(rollbackBuffer, rollbackBuffer.length << 1);
+      } else {
+        // hit threshold; not going to increase the buffer size
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private int pushRollbackBuffer(int idx, int bufferLength) {
+    if (idx >= bufferLength) {
+      return 0;
+    } else {
+      // note: we have already ensured sufficient buffer capacity
+      final char candidate = buffer.charAt(idx);
+      if (!UTF16.isLeadSurrogate(candidate)) {
+        if (rollbackBuffer != null) {
+          rollbackBuffer[rollbackBufferSize++] = candidate;
+        }
+        return 1;
+      } else {
+        if (++idx < bufferLength) {
+          if (rollbackBuffer != null) {
+            rollbackBuffer[rollbackBufferSize++] = candidate;
+            rollbackBuffer[rollbackBufferSize++] = buffer.charAt(idx);
+          }
+          return 2;
+        } else {
+          // we don't yet have the high surrogate
+          if (inputFinished) {
+            // no more input available; proceed with the char we have; rollback not needed
+            return 1;
+          } else {
+            return 0; // wait for more input to be available
+          }
+        }
+      }
+    }
+  }
+
+  private int readFromIoNormalizeUptoBoundary() {
+    int nextCharLength = pushRollbackBuffer(position.limit, buffer.length());
+    if (nextCharLength == 0) {
+      return 0;
+    }
+    final int preCharCount = charCount;
+    int preStart = position.start;
+    int preLimit;
+    do {
+      position.limit += nextCharLength;
+      preLimit = position.limit;
+      position.contextLimit = preLimit;
+      transform.filteredTransliterate(replaceable, position, true);
+      boolean rollbackSizeWithinBounds = true;
+      if (rollbackBuffer != null && position.start < position.limit && (rollbackSizeWithinBounds = ensureRollbackBufferCapacity())) {
+        // complete pass not transliterated, and not yet at rollback buffer threshold cap.
+        // N.b.: rollback buffer threshold cap is arbitrary, so we check here (somewhat
+        // counterintuitively, *before* rolling back, as opposed to before pushing to
+        // the buffer) so that we can fall through to the else clause, clear the rollback
+        // buffer, and proceed.
+        rollback(preStart, preLimit);
+      } else {
+        rollbackBufferSize = 0;
+        if (position.start > preStart) {
+          // cursor advanced
+          cursorAdvanced(preStart, preLimit);
+
+          // N.b.: advancing contextStart precludes support for quantifiers, but is crucial for streaming,
```

Comment:
I'm not sure. I went digging in the ICU code and couldn't find much at all about quantifiers. There's no obvious API for this, and the only clue I have is that ICU Transliterator currently [doesn't change `contextStart` at all](https://github.com/unicode-org/icu/blob/a075ac9c/icu4j/main/classes/translit/src/com/ibm/icu/text/Transliterator.java#L877), which if we did the same would result in reading the entire input into the `StringBuffer`, which @rmuir [warned to avoid](https://issues.apache.org/jira/browse/LUCENE-8972?focusedCommentId=16926589#comment-16926589). We could probably choose some arbitrary "maximum backup" to pad the `contextStart` and call it a day. This would support streaming processing but also probably satisfy "reasonable" uses of quantifiers. Does that sound ok?